In [5]:
import pandas as pd
import numpy as np

from pydataset import data

In [2]:
# data('mpg', show_doc=True) # view the documentation for the dataset
mpg = data('mpg') # load the dataset and store it in a variable


# 1. Copy the code from the lesson to create a dataframe full of student grades.

Create a column named passing_english that indicates whether each student has a passing grade in english.

In [40]:
import pandas as pd
import numpy as np

np.random.seed(123)

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# randomly generate scores for each student for each subject
# note that all the values need to have the same length here
math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))

df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades})

type(df)


pandas.core.frame.DataFrame

In [41]:
df

name  math  english  reading
0     Sally    62       85       80
1      Jane    88       79       67
2     Suzie    94       74       95
3     Billy    98       96       88
4       Ada    77       92       98
5      John    79       76       93
6    Thomas    82       64       81
7     Marie    93       63       90
8    Albert    92       62       87
9   Richard    69       80       94
10    Isaac    92       99       93
11     Alan    92       62       72

Sort the english grades by the passing_english column. How are duplicates handled?

In [9]:
df.sort_values(by= 'english')

name  math  english  reading
8    Albert    92       62       87
11     Alan    92       62       72
7     Marie    93       63       90
6    Thomas    82       64       81
2     Suzie    94       74       95
5      John    79       76       93
1      Jane    88       79       67
9   Richard    69       80       94
0     Sally    62       85       80
4       Ada    77       92       98
3     Billy    98       96       88
10    Isaac    92       99       93

Sort the english grades first by passing_english and then by student name. All the students that are failing english should be first, and within the students that are failing english they should be ordered alphabetically. The same should be true for the students passing english. (Hint: you can pass a list to the .sort_values method)

In [51]:
df['passing english'] = (df['english'] >= 70).astype(int)


df.sort_values(by=['passing english', 'name'],ascending= [True, True])

name  math  english  reading  passing english
11     Alan    92       62       72                0
8    Albert    92       62       87                0
7     Marie    93       63       90                0
6    Thomas    82       64       81                0
4       Ada    77       92       98                1
3     Billy    98       96       88                1
10    Isaac    92       99       93                1
1      Jane    88       79       67                1
5      John    79       76       93                1
9   Richard    69       80       94                1
0     Sally    62       85       80                1
2     Suzie    94       74       95                1

Sort the english grades first by passing_english, and then by the actual english grade, similar to how we did in the last step.

In [52]:
df.sort_values(by=['passing english', 'english'],ascending= [True, True])

name  math  english  reading  passing english
8    Albert    92       62       87                0
11     Alan    92       62       72                0
7     Marie    93       63       90                0
6    Thomas    82       64       81                0
2     Suzie    94       74       95                1
5      John    79       76       93                1
1      Jane    88       79       67                1
9   Richard    69       80       94                1
0     Sally    62       85       80                1
4       Ada    77       92       98                1
3     Billy    98       96       88                1
10    Isaac    92       99       93                1

Calculate each students overall grade and add it as a column on the dataframe. The overall grade is the average of the math, english, and reading grades.

In [55]:
df['overall grade'] = df[['math', 'english', 'reading']].mean(axis=1)

df

name  math  english  reading  passing english  overall grade
0     Sally    62       85       80                1      75.666667
1      Jane    88       79       67                1      78.000000
2     Suzie    94       74       95                1      87.666667
3     Billy    98       96       88                1      94.000000
4       Ada    77       92       98                1      89.000000
5      John    79       76       93                1      82.666667
6    Thomas    82       64       81                0      75.666667
7     Marie    93       63       90                0      82.000000
8    Albert    92       62       87                0      80.333333
9   Richard    69       80       94                1      81.000000
10    Isaac    92       99       93                1      94.666667
11     Alan    92       62       72                0      75.333333

# 2. Load the mpg dataset. Read the documentation for the dataset and use it for the following questions:

In [64]:
data('mpg', show_doc=True) 

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




In [63]:
mpg = data('mpg') 
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

How many rows and columns are there?

234 rows × 11 columns

What are the data types of each column?

In [65]:
mpg.dtypes

manufacturer     object
model            object
displ           float64
year              int64
cyl               int64
trans            object
drv              object
cty               int64
hwy               int64
fl               object
class            object
dtype: object

Summarize the dataframe with .info and .describe

In [75]:
mpg.info

mpg.describe



<bound method NDFrame.describe of     manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    

Rename the cty column to city.
Rename the hwy column to highway.

In [77]:
mpg = mpg.rename(columns = {'cty': 'city',
                          'hwy' : 'highway'})

mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

Do any cars have better city mileage than highway mileage?

In [133]:
mpg.sort_values(by='city', ascending=False).sort_values('highway', ascending=False) 





manufacturer                model  displ  year  cyl       trans drv  city  \
222   volkswagen           new beetle    1.9  1999    4  manual(m5)   f    35   
213   volkswagen                jetta    1.9  1999    4  manual(m5)   f    33   
223   volkswagen           new beetle    1.9  1999    4    auto(l4)   f    29   
197       toyota              corolla    1.8  2008    4  manual(m5)   f    28   
106        honda                civic    1.8  2008    4    auto(l5)   f    25   
..           ...                  ...    ...   ...  ...         ...  ..   ...   
70         dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   
66         dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4     9   
55         dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4     9   
60         dodge          durango 4wd    4.7  2008    8    auto(l5)   4     9   
127         jeep   grand cherokee 4wd    4.7  2008    8    auto(l5)   4     9   

     highway fl       class  mileage difference  
222       44  d  subcompact                   9  
213       44  d     compact                  11  
223       41  d  subcompact                  12  
197       37  r     compact                   9  
106       36  r  subcompact                  11  
..       ... ..         ...                 ...  
70        12  e      pickup                   3  
66        12  e      pickup                   3  
55        12  e      pickup                   3  
60        12  e         suv                   3  
127       12  e         suv                   3  

[234 rows x 12 columns]

no

In [98]:
mpg['mileage difference'] = mpg['highway'] - mpg['city']

mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage difference  
1    compact                  11  
2    compact                   8  
3    compact                  11  
4    compact                   9  
5    compact                  10  
..       ...                 ...  
230  midsize                   9  
231  midsize                   8  
232  midsize                  10  
233  midsize                   8  
234  midsize                   9  

[234 rows x 12 columns]

Which car (or cars) has the highest mileage difference?


In [132]:
mpg.sort_values(by='mileage difference', ascending=True).tail(2)

manufacturer       model  displ  year  cyl     trans drv  city  highway  \
223   volkswagen  new beetle    1.9  1999    4  auto(l4)   f    29       41   
107        honda       civic    1.8  2008    4  auto(l5)   f    24       36   

    fl       class  mileage difference  
223  d  subcompact                  12  
107  c  subcompact                  12

Which compact class car has the lowest highway mileage? The best?

In [129]:
pd.DataFrame(mpg[(mpg['class'] == 'compact')]).sort_values(by=['class', 'highway'], ascending=[True, True]).head(1)



manufacturer  model  displ  year  cyl     trans drv  city  highway fl  \
220   volkswagen  jetta    2.8  1999    6  auto(l4)   f    16       23  r   

       class  mileage difference  
220  compact                   7

In [131]:
 pd.DataFrame(mpg[(mpg['class'] == 'compact')]).sort_values(by=['class', 'highway'], ascending=[True, True])

manufacturer  model  displ  year  cyl       trans drv  city  highway fl  \
213   volkswagen  jetta    1.9  1999    4  manual(m5)   f    33       44  d   

       class  mileage difference  
213  compact                  11

Create a column named average_mileage that is the mean of the city and highway mileage.

In [163]:
mpg['average milage'] = mpg[['city', 'highway']].mean(axis=1)

mpg



manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage difference  average milage  
1    compact                  11            23.5  
2    compact                   8            25.0  
3    compact                  11            25.5  
4    compact                   9            25.5  
5    compact                  10            21.0  
..       ...                 ...             ...  
230  midsize                   9            23.5  
231  midsize                   8            25.0  
232  midsize                  10            21.0  
233  midsize                   8            22.0  
234  midsize                   9            21.5  

[234 rows x 13 columns]

Which dodge car has the best average mileage? The worst?

In [166]:
pd.DataFrame(mpg[(mpg['manufacturer'] == 'dodge')]).sort_values(by='average milage', ascending = True).tail(1)

manufacturer        model  displ  year  cyl     trans drv  city  highway  \
38        dodge  caravan 2wd    2.4  1999    4  auto(l3)   f    18       24   

   fl    class  mileage difference  average milage  
38  r  minivan                   6            21.0

In [167]:

pd.DataFrame(mpg[(mpg['manufacturer'] == 'dodge')]).sort_values(by='average milage', ascending = True).head(2)

manufacturer                model  displ  year  cyl       trans drv  city  \
70        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   
66        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4     9   

    highway fl   class  mileage difference  average milage  
70       12  e  pickup                   3            10.5  
66       12  e  pickup                   3            10.5

# Load the mpg dataset. Read the documentation for the dataset and use it for the following questions:

In [188]:
data('Mammals', show_doc=True)
mammals=data('Mammals')


Mammals

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Garland(1983) Data on Running Speed of Mammals

### Description

Observations on the maximal running speed of mammal species and their body
mass.

### Usage

    data(Mammals)

### Format

A data frame with 107 observations on the following 4 variables.

weight

Body mass in Kg for "typical adult sizes"

speed

Maximal running speed (fastest sprint velocity on record)

hoppers

logical variable indicating animals that ambulate by hopping, e.g. kangaroos

specials

logical variable indicating special animals with "lifestyles in which speed
does not figure as an important factor": Hippopotamus, raccoon (Procyon),
badger (Meles), coati (Nasua), skunk (Mephitis), man (Homo), porcupine
(Erithizon), oppossum (didelphis), and sloth (Bradypus)

### Details

Used by Chappell (1989) and Koenker, Ng and Portnoy (1994) to illustrate the
fitting of piecewise linear curves.

### Source

Garland, T. (

In [173]:
mammals

weight  speed  hoppers  specials
1    6000.000   35.0    False     False
2    4000.000   26.0    False     False
3    3000.000   25.0    False     False
4    1400.000   45.0    False     False
5     400.000   70.0    False     False
6     350.000   70.0    False     False
7     300.000   64.0    False     False
8     260.000   70.0    False     False
9     250.000   40.0    False     False
10   3800.000   25.0    False      True
11   1000.000   60.0    False     False
12    900.000   70.0    False     False
13    900.000   56.0    False     False
14    800.000   29.0    False     False
15    750.000   57.0    False     False
16    500.000   32.0    False     False
17    450.000   56.0    False     False
18    300.000   72.0    False     False
19    300.000   90.0    False     False
20    250.000   80.0    False     False
21    250.000   56.0    False     False
22    170.000   80.0    False     False
23    150.000   48.0    False     False
24    130.000   70.0    False     False
25    120.000   80.0    False     False
26    120.000   61.0    False     False
27    110.000   33.0    False     False
28    100.000   64.0    False     False
29     85.000   55.0    False     False
30     80.000   65.0    False     False
31     72.000   56.0    False     False
32     70.000   45.0    False     False
33     65.000   60.0    False     False
34     62.000   81.0    False     False
35     50.000  100.0    False     False
36     50.000   60.0    False     False
37     50.000   40.0    False     False
38     50.000   47.0    False     False
39     37.000  105.0    False     False
40     35.000   80.0    False     False
41     34.000   97.0    False     False
42     30.000   97.0    False     False
43     30.000   80.0    False     False
44     30.000   45.0    False     False
45     20.000   81.0    False     False
46    400.000   40.0    False     False
47    300.000   48.0    False     False
48    230.000   56.0    False     False
49    150.000   59.0    False     False
50    135.000   48.0    False     False
51     65.000   65.0    False     False
52     60.000   60.0    False     False
53     55.000  110.0    False     False
54     45.000   50.0    False     False
55     40.000   64.0    False     False
56     25.000   67.0    False     False
57     20.000   70.0    False     False
58     16.000   65.0    False     False
59     12.000   24.0    False      True
60     11.000   30.0    False      True
61     10.000   56.0    False     False
62      7.000   60.0    False     False
63      6.000   72.0    False     False
64      5.000   64.0    False     False
65      5.000   27.0    False      True
66      3.000   16.0    False      True
67    127.000   32.0    False     False
68     70.000   40.0    False      True
69     13.000   37.0    False      True
70      9.000    3.2    False      True
71      4.000   16.0    False     False
72      0.600   36.0    False     False
73      0.600   20.0    False     False
74      0.550   27.0    False     False
75      0.500   18.0    False     False
76      0.400   20.0    False     False
77      0.300   13.0    False     False
78      0.250    9.7    False     False
79      0.220   15.0    False     False
80      0.110    9.0    False     False
81      0.100   17.0    False     False
82      0.056   21.0     True     False
83      0.050   11.0    False     False
84      0.045   16.0    False     False
85      0.035   32.0     True     False
86      0.035   14.0     True     False
87      0.030    6.8    False     False
88      0.030    9.1    False     False
89      0.025   11.0    False     False
90      0.025    8.6    False     False
91      0.018    8.9    False     False
92      0.016   13.0    False     False
93      0.100    4.0    False     False
94      0.100    2.4    False     False
95      0.016    3.6    False     False
96      4.600   64.0     True     False
97      4.400   72.0     True     False
98      4.000   72.0     True     False
99      3.500   56.0     True     False
100    

How many rows and columns are there? 

In [189]:
mammals.shape

(107, 4)

What are the data types?

In [177]:
mammals.dtypes

weight      float64
speed       float64
hoppers        bool
specials       bool
dtype: object

Summarize the dataframe with .info and .describe

In [190]:
mammals.info()

<class 'pandas.core.frame.DataFrame'>
Index: 107 entries, 1 to 107
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   weight    107 non-null    float64
 1   speed     107 non-null    float64
 2   hoppers   107 non-null    bool   
 3   specials  107 non-null    bool   
dtypes: bool(2), float64(2)
memory usage: 2.7 KB


In [194]:
mammals.describe(include='all')

weight       speed hoppers specials
count    107.000000  107.000000     107      107
unique          NaN         NaN       2        2
top             NaN         NaN   False    False
freq            NaN         NaN      96       97
mean     278.688178   46.208411     NaN      NaN
std      839.608269   26.716778     NaN      NaN
min        0.016000    1.600000     NaN      NaN
25%        1.700000   22.500000     NaN      NaN
50%       34.000000   48.000000     NaN      NaN
75%      142.500000   65.000000     NaN      NaN
max     6000.000000  110.000000     NaN      NaN

What is the the weight of the fastest animal?

In [196]:
mammals.sort_values('speed', ascending=False).head(1).weight

53    55.0
Name: weight, dtype: float64

What is the overal percentage of specials?

In [197]:
10/107

0.09345794392523364

In [200]:
mammals.specials.mean()

0.09345794392523364

In [201]:
round(mammals.specials.mean() * 100, 2)

9.35

How many animals are hoppers that are above the median speed? What percentage is this?

In [202]:
mammals.speed.median()

48.0

In [203]:
hoppers = mammals[mammals.hoppers]
hoppers
                  

weight  speed  hoppers  specials
82    0.056   21.0     True     False
85    0.035   32.0     True     False
86    0.035   14.0     True     False
96    4.600   64.0     True     False
97    4.400   72.0     True     False
98    4.000   72.0     True     False
99    3.500   56.0     True     False
100   2.000   64.0     True     False
101   1.900   56.0     True     False
102   1.500   50.0     True     False
103   1.500   40.0     True     False

In [204]:
numb_hoppers =((hoppers.speed) > (mammals.speed.median())).sum()
numb_hoppers

7

In [205]:
len(mammals)

107

In [207]:
numb_hoppers / len(mammals)

0.06542056074766354